### PR TITLE
feat: more informative error messages when Test Replay fails to capture

### LINKED ISSUE #29022: Test Replay: [Upload] Display more informative errors
In certain cases, the app can display cryptic error messages that are not very helpful - to our users, or to us.

Failed Upload error messages should display after the "Uploaded Cloud Artifacts" section.

Failed Recording error messages should display before the "Uploading Cloud Artifacts" section.

**Failed Uploads, Cause: HTTP Status Code 403, 404, 413**

Source issues:
* https://github.com/cypress-io/cypress-services/issues/7722 (403)
* https://github.com/cypress-io/cypress-services/issues/7732 (404)
* https://github.com/cypress-io/cypress-services/issues/7723 (413)

Acceptance: surface more informative error:

> Unable to upload this spec's Test Replay recording due to an error during the upload process:
>
> $URL responded with HTTP $CODE, $DESCRIPTION

**Failed Uploads, Cause: Network Error**
Socket Hangup, Proxy Error, ECONNRESET are likely caused by a network or network configuration error.
Acceptance: surface a more informative error; solution: fix network configuration

Acceptance: surface a more informative error:

> Unable to upload this spec's Test Replay recording due to a network error during the upload process: $ERR
>
> Please verify your network configuration for accessing $URL. 

Source issues:
* https://github.com/cypress-io/cypress/issues/28004
* https://github.com/cypress-io/cypress-services/issues/7731
* https://github.com/cypress-io/cypress-services/issues/7566

**Failed Uploads, Cause: runtime error**
Sometimes we encounter an "cannot read .pipes of undefined" error during upload. This is being thrown from ReadableStream with an unknown cause. 
Acceptance: Display an error:

> Unable to upload this spec's Test Replay recording. An error occurred while trying to read $FILENAME:
>
> $ERR

**Failed Initialization**
Sometimes, we fail to initialize Protocol in a way that does not throw a terminal error. This results in either an undefined `archivePath` on `ProtocolManager.ts` in `app` (`Error: no test data recorded`). Sometimes, an inability to open the sqlite file in `beforeSpec` results on a terminal error (`SqliteError: Unable to open database file`). Messaging on both can be improved:

> Unable to initialize Test Replay recording for this spec (TBD - determine what is failing, and throw an appropriate error).

> Unable to initialize Test Replay recording for this spec. (TBD: should we display the path to the sqlite file we were trying to open?)

**Failed Recording**
An `unknown error, ftruncate` error seems to be thrown from `indexed-tarfile` when we append, sometimes.
Acceptance: Display a more informative error; potential solution: increase available disk space
> Unable to add one or more Test Replay assets to the recording. Please verify disk space availability.

`SqliteError: Database or Disk is Full` can be encountered whenever we write to the database, or close it. 
Acceptance: Display a more informative error; solution: increase available disk space
> Unable to add one or more Test Replay events to the recording. Please verify disk space availability.

### PR DESCRIPTION
**Pull request by cacieprins**
- Closes https://github.com/cypress-io/cypress/issues/29022

### Additional details
Many of the errors that can crop up when initializing, capturing, or uploading Test Replay recordings were fairly cryptic. This PR attempts to improve that, with differentiated error messages for:
* Test Replay recordings failing to initialize or capture
* Test Replay recordings failing to upload due to system-level network errors
* Test Replay recordings failing to upload due to HTTP errors returned from the archive upload endpoint

In addition, Test Replay will cease to display in the Artifact Upload manifest if it fails to initialize or if it encounters a fatal error during recording. We print an error message before the upload section of stdout in this case.

Several of these more verbose error messages include paths and URLs that will more readily identify issues and lead to faster and easier resolution. In certain situations, we can provide actionable advice, like: confirming network configuration, increasing available disk space, and ensuring that temporary directories are accessible.

### Steps to test
* Run record_spec system tests
* Simulate network failures specific to s3 while running a spec that records to Test Replay
* Simulate HTTP errors specific to s3 while running a spec that records to Test Replay

### How has the user experience changed?
The existing upload report will still be displayed, however, the following error messages will also be displayed:

#### When a request to our API to confirm the status of artifact (video, screenshots, and Test Replay) fails
```
Warning: We encountered an error while confirming the upload of artifacts for this spec.

These results will not display artifacts.

This error will not affect or change the exit code.

StatusCodeError: 500 - "Internal Server Error"
```
#### When Test Replay fails to initialize

```
Warning: we encountered an error while initializing the Test Replay recording for this spec.
        
These results will not display Test Replay recordings.
        
This error will not affect or change the exit code.

(Error message and stack trace)
```

#### When Test Replay encounters a fatal  error during recording
```
Warning: We encountered an error while recording Test Replay data for this spec.
        
These results will not display Test Replay recordings.

This can happen for many reasons. If this problem persists:

- Try increasing the available disk space.
- Ensure that /os/temporary-directory/cypress/protocol s both readable and writable.

This error will not affect or change the exit code.

(Error message and stack trace)
```

#### When S3 returns an HTTP status code that indicates an error
```
Warning: We encountered an HTTP error while uploading the Test Replay recording for this spec.

These results will not display Test Replay recordings.

This error will not affect or change the exit code.

[URL] responded with HTTP 500: Internal Server Error
```

#### When the Test Replay recording fails to upload due to a system-level network error
```
Warning: We encountered a network error while uploading the Test Replay recording for this spec.

Please verify your network configuration for accessing https://some/url

These results will not display Test Replay recordings.

This error will not affect or change the exit code.

Error: request to http://fake.test/url failed, reason: getaddrinfo ENOTFOUND fake.test
```

#### When the Test Replay recording fails to upload after several retries
An aggregate error is displayed with each error that was encountered while uploading. We display a recommendation to verify network configuration when we detect that at least one of these errors is a system-level network error.

```
Warning: We encountered multiple errors while uploading the Test Replay recording for this spec.

We attempted to upload the Test Replay recording 3 times.

Some or all of the errors encountered are system-level network errors. Please verify your network configuration for connecting to http://some/url

In addition to the stack trace provided, 2 errors were encountered. Stack traces for these errors are omitted for brevity:
 - (error message)
 - (error message)

Error: http://some/url: ECONNRESET
    [STACK]
```

### PR Tasks

- [X] Have tests been added/updated?
- [NA] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [NA] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Review by jennifer-shehane — COMMENTED** [review]
I like the error message text that will be displayed. I wish they could be a little more actionable, but I don't quite see how they can be. 

If we had any expectations of wanting to change the actions we recommend to people in this situations in the future, I would recommend putting an on.cypress.io link to our error-messages page so that we could more easily update the actions. But I'm not sure this is the case with these situations.

File: packages/server/lib/cloud/artifacts/upload_artifacts.ts [hunk 1]
```diff
@@ -151,23 +207,42 @@ export const uploadArtifacts = async (options: UploadArtifactOptions) => {
...
+        console.log('')
+        errors.warning('CLOUD_PROTOCOL_UPLOAD_NEWORK_FAILURE', error)
+      }
+    }
 
-    /**
-     * Protocol instances with fatal errors prior to uploading will not have an uploadResult,
-     * but we still want to report them to updateInstanceArtifacts
-     */
+    // there are no upload results entry for protocol if we did not attempt to upload protocol due to a fatal error
```
Comment (ryanthemanuel) [inline packages/server/lib/cloud/artifacts/upload_artifacts.ts]:
```suggestion
    // there is no upload results entry for protocol if we did not attempt to upload protocol due to a fatal error
```

File: packages/server/lib/cloud/api/scrub_url.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+const SENSITIVE_KEYS = Object.freeze(['x-amz-credential', 'x-amz-signature', 'Signature', 'AWSAccessKeyId'])
+
+export const scrubUrl = (url: string): string => {
+  const parsedUrl = new URL(url)
+
+  for (const [key, value] of parsedUrl.searchParams) {
+    if (SENSITIVE_KEYS.includes(key)) {
+      parsedUrl.searchParams.set(key, 'X'.repeat(value.length))
```
Comment (ryanthemanuel) [inline packages/server/lib/cloud/api/scrub_url.ts:8]: Do we want to match the value length? Not sure if just the length of the string could give something away?

File: packages/server/lib/cloud/artifacts/upload_artifacts.ts [hunk 3]
```diff
@@ -102,9 +108,23 @@ const extractArtifactsFromOptions = async ({
 
     const protocolUploadUrl = captureUploadUrl || protocolCaptureMeta.url
 
-    debug('should add protocol artifact? %o, %o, %O', protocolFilePath, protocolUploadUrl, protocolManager)
-    if (protocolManager && protocolFilePath && protocolUploadUrl) {
-      artifacts.push(await createProtocolArtifact(protocolFilePath, protocolUploadUrl, protocolManager))
+    const shouldAddProtocolArtifact = protocolManager && protocolFilePath && protocolUploadUrl && !protocolManager.hasFatalError()
+
+    debug('should add protocol artifact? %o, %o, %O', {
```
Comment (ryanthemanuel) [inline packages/server/lib/cloud/artifacts/upload_artifacts.ts]: Should this be %o %o %O? It seems like there's only one object here.

**Review by ryanthemanuel — APPROVED** [review]
Looks really good. A couple of smaller things.

File: packages/server/lib/cloud/artifacts/upload_artifacts.ts (hunk 3 above)
Comment (cacieprins) [reply packages/server/lib/cloud/artifacts/upload_artifacts.ts]: Good catch - there used to be more

File: packages/server/lib/cloud/api/scrub_url.ts (hunk 2 above)
Comment (cacieprins) [reply packages/server/lib/cloud/api/scrub_url.ts:8]: these keys are fairly temporary, so I'm not sure if it's even necessary to be honest.

File: cli/CHANGELOG.md [hunk 4]
```diff
@@ -1,12 +1,16 @@
 <!-- See the ../guides/writing-the-cypress-changelog.md for details on writing the changelog. -->
-## 13.8.1
+## 13.9.0
```
Comment (mschile) [inline cli/CHANGELOG.md:2]: There should be a new section for `13.9.0` since `13.8.1` has already been released.

**Review by mschile — APPROVED** [review]
Looks good! One small comment on the changelog.

File: cli/CHANGELOG.md (hunk 4 above)
Comment (cacieprins) [reply cli/CHANGELOG.md:2]: addressed in https://github.com/cypress-io/cypress/pull/29312/commits/122cd50cf0a3aae0418874dba7b6146097866532

File: cli/CHANGELOG.md [hunk 5]
```diff
@@ -1,6 +1,8 @@
 <!-- See the ../guides/writing-the-cypress-changelog.md for details on writing the changelog. -->
 ## 13.9.0
 
+_Release 5/7/2024 (PENDING)_
```
Comment (jennifer-shehane) [inline cli/CHANGELOG.md:4]: @cacieprins typo

```
_Released 5/7/2024 (PENDING)_
```

File: cli/CHANGELOG.md [hunk 6]
```diff
@@ -1,4 +1,12 @@
 <!-- See the ../guides/writing-the-cypress-changelog.md for details on writing the changelog. -->
+## 13.9.0
+
+_Release 5/7/2024 (PENDING)_
```
Comment (jennifer-shehane) [inline cli/CHANGELOG.md]: @cacieprins Typo

```suggestion
_Released 5/7/2024 (PENDING)_
```

**Pull request merged**